plugins/cilium-cni/lib/deletion_queue.go

```diff
@@ -0,0 +1,133 @@
+
+	lockfile *lockfile.Lockfile
+}
+
+// the timeout for connecting and obtaining the lock
+// the default of 30 seconds is too long; kubelet will time us out before then
+const timeoutSeconds = 10
+
+// NewDeletionFallbackClient creates a client that will either issue an EndpointDelete
+// request via the api, *or* queue one in a temporary directory.
+// To prevent race conditions, the logic is:
+// 1. Try and connect to the socket. if that succeeds, done
+// 2. Otherwise, take a shared lock on the delete queue directory
+// 3. Once we get the lock, check to see if the socket now exists
+// 4. If it exists, drop the lock and use the api
+func NewDeletionFallbackClient(logger *logrus.Entry) (*DeletionFallbackClient, error) {
+	dc := &DeletionFallbackClient{
+		logger: logger,
+	}
+
+	// Try and connect (the usual case)
+	err := dc.tryConnect()
+	if err == nil {
+		return dc, nil
+	}
+	dc.logger.WithError(err).Warnf("Failed to connect to agent socket at %s.", client.DefaultSockPath())
+
+	// We failed to connect: get the queue lock
+	if err := dc.tryQueueLock(); err != nil {
+		return nil, fmt.Errorf("failed to acquire deletion queue: %w", err)
+	}
+
+	// We have the queue lock; try and connect again
+	// just in case the agent finished starting up while we were waiting
+	if err := dc.tryConnect(); err == nil {
+		dc.logger.Info("Successfully connected to API on second try.")
+		// hey, it's back up!
+		dc.lockfile.Unlock()
+		dc.lockfile = nil
+		return dc, nil
+	}
+
+	// We have the lockfile, but no valid client
+	dc.logger.Info("Agent is down, falling back to deletion queue directory")
+	return dc, nil
+}
+
+func (dc *DeletionFallbackClient) tryConnect() error {
+	c, err := client.NewDefaultClientWithTimeout(timeoutSeconds * time.Second)
+	if err != nil {
+		return err
+	}
+	dc.cli = c
+	return nil
+}
+
+func (dc *DeletionFallbackClient) tryQueueLock() error {
+	dc.logger.Debugf("attempting to acquire deletion queue lock at %s", defaults.DeleteQueueLockfile)
+
+	// Ensure deletion queue directory exists, obtain shared lock
+	err := os.MkdirAll(defaults.DeleteQueueDir, 0755)
+	if err != nil {
+		return fmt.Errorf("failed to create deletion queue directory %s: %w", defaults.DeleteQueueDir, err)
+	}
+
+	lf, err := lockfile.NewLockfile(defaults.DeleteQueueLockfile)
+	if err != nil {
+		return fmt.Errorf("failed to open lockfile %s: %w", defaults.DeleteQueueLockfile, err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutSeconds*time.Second)
+	defer cancel()
+
+	err = lf.Lock(ctx, false) // get the shared lock
+	if err != nil {
+		return fmt.Errorf("failed to acquire lock: %w", err)
+	}
+	dc.lockfile = lf
+	return nil
+}
+
+// EndpointDelete deletes an endpoint given by an endpoint id, either
+// by directly accessing the API or dropping in a queued-deletion file.
+// endpoint-id is a qualified endpoint reference, e.g. "container-id:XXXXXXX"
+func (dc *DeletionFallbackClient) EndpointDelete(id string) error {
```

Comment:
Should there be a cap on how many queued deletes we allow? Just in case it grows indefinitely?

Reply:
That's a good idea for a few reasons
- protect against some kind of kubelet total insanity
- prevent against a downgrade where the agent is no longer processing deletions

I'll add it.

Reply:
Done!

Reply:
I agree this is a nice change. I wonder if it would also be useful if we had some sort of preemptive warning as a log msg as we approach the hard limit (maybe at 75%, 90%, 95%?). Given the CNI binary has no way to export metrics AFAIK, we don't really have visibility as to what's happening with this queue. Also open to other ideas that could give us visibility. 

In general, one of the major goals for Cilium moving forward is better supportability and that includes visibility. One quick way to do that is to enumerate the hard limits in Cilium and ensure Cilium doesn't just all of a sudden stop to a grinding halt when it reaches them.

Reply:
That's a fair observation, and I also dislike systems that have surprises. In this case, it's difficult to export this as a metric because it's only in play when the agent is down :-).

The potential failure mode here - the queue filling up - is equivalent to the behavior before this PR: CNI deletion fails explicitly and is retried by the kubelet. Very different semantics than, say, bpf map exhaustion. Given that there is a clean fallback, I'd say it's OK not to add complicated guard-rails or alerting here.

Reply:
Fair enough. A note in the code where we fallback would be good to call out as a comment so that future readers know that that was an explicit design decision. We can also revisit it in the future if this theoretical problem actually turns up in the wild and at least we will have documented our assumptions of it in the code.